conda_declarative/state.py

```diff
@@ -0,0 +1,217 @@
+
+    Note that not all fields of the `Environment` model are supported
+    here.
+
+    Parameters
+    ----------
+    prefix : str | pathlib.Path
+        Prefix of the environment
+
+    Returns
+    -------
+    Environment | None
+        The Environment model, if the env file exists; None otherwise
+    """
+    env_path = get_env_path(prefix)
+    if env_path.exists():
+        with open(env_path) as f:
+            env_dict = yaml_safe_load(f.read())
+
+        # Coerce any non-primitive types to primitive types so they
+        # can be (de)serialized
+        env_dict["requested_packages"] = list(
+            map(MatchSpec, env_dict.get("requested_packages", []))
+        )
+
+        if "config" in env_dict:
+            config = env_dict["config"]
+
+            if "aggressive_update_packages" in config and isinstance(
+                config["aggressive_update_packages"], tuple | list
+            ):
+                config["aggressive_update_packages"] = tuple(
+                    map(MatchSpec, config["aggressive_update_packages"])
+                )
+
+            # Convert the string values to their enum type counterparts
+            config["channel_priority"] = ChannelPriority(
+                config.get("channel_priority", ChannelPriority.FLEXIBLE)
+            )
+            config["deps_modifier"] = DepsModifier(
+                config.get("deps_modifier", DepsModifier.NOT_SET)
+            )
+            config["sat_solver"] = SatSolverChoice(
+                config.get("sat_solver", SatSolverChoice.PYCOSAT)
+            )
+            config["update_modifier"] = UpdateModifier(
+                config.get("update_modifier", UpdateModifier.UPDATE_SPECS)
+            )
+
+            env_dict["config"] = EnvironmentConfig(**config)
+        else:
+            env_dict["config"] = None
+
+        return Environment(**env_dict)
+
+    return None
+
+
+def to_env_file(environment: Environment):
+    """Write the Environment to the appropriate path in the environment directory.
+
+    Note that not all fields of the `Environment` model are supported. Fields that are
+    either primitive types or datacalsses, or primitive containers of primitive types
+    or dataclasses are handled automatically. Other fields are not, with the exception
+    of the `requested_packages`, which is the only non-primitive field used here.
+
+    Parameters
+    ----------
+    environment : Environment
+        Environment model to serialize and write to disk
+    """
+    env_dict = asdict(environment)
+
+    # Coerce any non-primitive types to primitive types so they
+    # can be (de)serialized
+    env_dict["requested_packages"] = list(
+        map(str, env_dict.get("requested_packages", []))
+    )
+
+    config = env_dict["config"]
+
+    # aggressive_update_packages can either be a bool or a tuple[MatchSpec]
+    if "aggressive_update_packages" in config and isinstance(
+        config["aggressive_update_packages"], tuple | list
```

Comment:
Huh, can we use unions there now? I thought we needed this:

```suggestion
        config["aggressive_update_packages"], (tuple, list)
```